settings/src/components/password.js

```diff
@@ -1,21 +1,72 @@
+/**
+ * External dependencies
+ */
+import { pick } from 'lodash';
+// window.zxcvbn is also used, but can't be declared here because it's loaded asyncronously.
+
 /**
  * WordPress dependencies
  */
-import { Button, TextControl } from '@wordpress/components';
+import { Button, Flex, Notice, TextControl }     from '@wordpress/components';
+import { useCallback, useEffect, useState }      from '@wordpress/element';
+import { Icon, cancelCircleFilled, check, seen } from '@wordpress/icons';
+
+import {} from '@wordpress/util'; // Enqueue via asset.php.
+const { post } = wp.ajax;
 
 /**
  * Render the Password setting.
  */
-export default function Password() {
-	// todo
-	// use same generator as core, do server side?
-	// strength meter
-	// warn when weak
-	// automatically generate placeholder?
-	// how to get empty value so placeholder shows up?
+export default function Password( { userRecord, userRequires2fa } ) {
+	const [ passwordStrong, setPasswordStrong ] = useState( false );
+	const [ saved, setSaved ]                   = useState( false );
+	const [ generating, setGenerating ]         = useState( false );
+	const [ inputType, setInputType ]           = useState( 'password' );
+
+	useEffect( () => {
+		if ( ! userRecord.editedRecord.password ) {
+			return;
+		}
+
+		setSaved( false );
+		setPasswordStrong( isPasswordStrong( userRecord.editedRecord.password, userRecord.record, userRequires2fa ) );
+	}, [ userRecord.editedRecord.password ] );
+
+	/**
+	 * @todo When starting from a fresh page load, theres a flash of the red "too easy" notice before it shows
+	 * the success notice. It happens between the time when `edit()` is called and the `useEffect` callback fires
+	 * to update the `passwordStrong` state. Is there a way to tell React to wait until both are done to re-render?
+	 * Or maybe the condition that renders the notice can include something like `hasResolved`?
+	 */
+	const generatePassword = useCallback( async () => {
+		setGenerating( true );
+		const password = await post( 'generate-password' );
+
+		userRecord.edit( { password } );
+		setInputType( 'text' );
+		setGenerating( false );
+	}, [] );
+
+	/**
+	 * @todo This works the first time it's called, but subsequent calls result in a `403` error. That might be
+	 * because changing the password resets the auth cookies. The failed request is automatically repeated by
+	 * `useEntityRecord.save()`, though, and that request succeeds. The only tangible problem here is the console
+	 * error, but it'd be nice to fix that.
+	 */
```

Comment:
I'm not seeing this behaviour, the `POST` completes with a `200` and returns the new user object.

What's the request/contents of the request for the 403 and subsequent 200?

Reply:
🤔 it happens for me in FF and Chrome.

https://user-images.githubusercontent.com/484068/210116102-002fda15-e6ed-4ad4-bb99-eb4bceb47218.mov

Reply:
fixed in 6ac3885

Reply:
Ahhhh this is an issue on local environments, but not on WordPress.org, as WordPress.org doesn't have the nonces tied to the login session.

This sounds like a bug though, The `X-WP-Nonce` header returned on the `POST wp-json...` endpoint should be updated to the new value, it might be that it's used for future REST API's but not for admin-ajax endpoints.. OR the `set password` code in the rest api is re-generating the nonce prior to the new password being set, that invalidates the existing session.

Reply:
> not [an issue] on WordPress.org, as WordPress.org doesn't have the nonces tied to the login session.

Ah, ok 👍🏻 

> it might be that it's used for future REST API's but not for admin-ajax endpoints

Were you thinking this was interacting w/ the `generate-password` AJAX endpoint (which was removed in https://github.com/WordPress/wporg-two-factor/commit/5241afeac49739ba58d8162f468434291dbc0fb6)? I can see those not being compatible, but that doesn't seem to be an issue here since all the requests in the `savePasswordHandler()` use `userRecord.save()`.

I'm guessing it's harmless to leave this on in production, but if you think it's best I could look into passing `wp_get_environment_type()` to JS, and wrapping this in a conditional.

Reply:
Nah Leave it in production, we'll need it eventually.

My point was that in the event of `user.save()` the new Authentication cookies are served through the rest api, but, the nonces aren't refreshed automatically (although the REST API responds with a new nonce as part of the `.save()` - see the response headers for the x-wp-nonce).

I haven't tested your scenario, I was spit-balling, I don't know which of the nonces are invalid (one, or both) or if it should be automatically updating it or not. My gut-feeling is that the response should be responding with a new ajax nonce on the rest api XHR potentially, but the presence of the `refreshNonce` API seems.. well I guess it does the trick.

Reply:
That sounds good 👍🏻